From b5c62ae463101712a2ed91e976b28af5d1e73064 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michel=20D=C3=A4nzer?= <mdaenzer@redhat.com>
Date: Fri, 6 Nov 2020 10:14:19 +0100
Subject: [PATCH xserver 25/25] xwayland: Make window_get_client_toplevel
 non-recursive

Noticed while reading the code.

Reviewed-by: Olivier Fourdan <ofourdan@redhat.com>
(cherry picked from commit df3aa4922fd7e256169e541188b724f67ca948e1)
---
 hw/xwayland/xwayland.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/hw/xwayland/xwayland.c b/hw/xwayland/xwayland.c
index 1600c00cd..a5756ea14 100644
--- a/hw/xwayland/xwayland.c
+++ b/hw/xwayland/xwayland.c
@@ -578,11 +578,11 @@ window_get_client_toplevel(WindowPtr window)
      * decoration/wrapper windows. In that case recurse by checking the client
      * of the first *and only* child of the decoration/wrapper window.
      */
-    if (window_is_wm_window(window)) {
-        if (window->firstChild && window->firstChild == window->lastChild)
-            return window_get_client_toplevel(window->firstChild);
-        else
+    while (window_is_wm_window(window)) {
+        if (!window->firstChild || window->firstChild != window->lastChild)
             return NULL; /* Should never happen, skip resolution emulation */
+
+        window = window->firstChild;
     }
 
     return window;
-- 
2.28.0

